tests/reprsimil/test_brsa.py

```diff
@@ -0,0 +1,211 @@
+    noise = np.zeros([n_T,n_V])
+    noise[0,:] = np.random.randn(n_V) * noise_level / np.sqrt(1-rho1**2)
+    for i_t in range(1,n_T):
+        noise[i_t,:] = noise[i_t-1,:] * rho1 +  np.random.randn(n_V) * noise_level
+    
+    # ideal covariance matrix
+    ideal_cov = np.zeros([n_C,n_C])
+    ideal_cov = np.eye(n_C)*0.6
+    ideal_cov[0,0] = 0.2
+    ideal_cov[5:9,5:9] = 0.6
+    for cond in range(5,9):
+        ideal_cov[cond,cond] = 1
+    idx = np.where(np.sum(np.abs(ideal_cov),axis=0)>0)[0]
+    L_full = np.linalg.cholesky(ideal_cov)        
+
+    # generating signal
+    snr_level = 5.0 # test with high SNR    
+    # snr = np.random.rand(n_V)*(snr_top-snr_bot)+snr_bot
+    # Notice that accurately speaking this is not snr. the magnitude of signal depends
+    # not only on beta but also on x.
+    inten = np.random.randn(n_V) * 20.0
+
+    # parameters of Gaussian process to generate pseuso SNR
+    tau = 0.8
+    smooth_width = 5.0
+    inten_kernel = 1.0
+    
+    coords = np.arange(0,n_V)[:,None]
+
+    dist2 = np.square(coords-coords.T)
+    
+    inten_tile = np.tile(inten,[n_V,1])
+    inten_diff2 = (inten_tile-inten_tile.T)**2
+
+    K = np.exp(-dist2/smooth_width**2/2.0 -inten_diff2/inten_kernel**2/2.0) * tau**2 + np.eye(n_V)*tau**2*0.001
+
+    L = np.linalg.cholesky(K)
+    snr = np.exp(np.dot(L,np.random.randn(n_V))) * snr_level
+    sqrt_v = noise_level*snr
+    betas_simulated = np.dot(L_full,np.random.randn(n_C,n_V)) * sqrt_v
+    signal = np.dot(design.design_used,betas_simulated)
+    
+    
+    Y = signal + noise
+
+
+    scan_onsets = np.linspace(0,design.n_TR,num=5)
+
+    # Testing with GP prior.
+    brsa = BRSA(GP_space=True,GP_inten=True,verbose=True,n_iter = 20)
+
+    brsa.fit(X=Y, design=design.design_used, scan_onsets=scan_onsets,
+             coords=coords, inten=inten)
+    
+    # Check that result is significantly correlated with the ideal covariance matrix
+    u_b = brsa.U_[1:,1:]
+    u_i = ideal_cov[1:,1:]
+    p = scipy.stats.spearmanr(u_b[np.tril_indices_from(u_b,k=-1)],u_i[np.tril_indices_from(u_i,k=-1)])[1]
+    assert p < 0.01, "Fitted covariance matrix does not correlate with ideal covariance matrix!"
+    # check that the recovered SNR makes sense
+    p = scipy.stats.pearsonr(brsa.nSNR_,snr)[1]
+    assert p < 0.01, "Fitted SNR does not correlate with simualted SNR!"
+    assert np.isclose(np.mean(np.log(brsa.nSNR_)),0), "nSNR_ not normalized!"
+
+    assert np.abs(brsa.bGP_ - tau) / tau < 0.3, "standard deviation of GP deviates too much"
+    assert np.abs(brsa.lGPspace_ - smooth_width) / smooth_width < 0.7,\
+        "spatial length scale of GP deviates too much"
+    assert np.abs(brsa.lGPinten_ - inten_kernel) / inten_kernel < 0.7,\
+        "intensity length scale of GP deviates too much"
+    
+    # test if gradient is correct
+    XTY,XTDY,XTFY,YTY_diag, YTDY_diag, YTFY_diag, XTX, XTDX, XTFX = brsa._prepare_data(design.design_used,Y,n_T,n_V,scan_onsets)
+    n_l = n_C*(n_C+1)/2
```

Comment:
Probably should do this using something like http://numdifftools.readthedocs.io/en/latest/ rather than simple first order (will also allow to shrink tolerance). Other alternative is to compare against a symbolic diff tool like theano but that's more work so maybe later.